docs/rate-limit-and-backpressure.md

```diff
@@ -0,0 +1,128 @@
+
+## Why can Barrier Latency be High?
+
+The data between barriers takes a long time to process.
+This can be due to high amplification, UDF latency, or other reasons.
+
+By default, Barriers have an interval of 1s.
+If records between those barrier take longer than 256ms to process, the barrier latency will increase by that difference.
+
+## How does Backpressure affect Barrier Latency?
+
+### A simple scenario without channels
+
+In our system, we have Actors which run various parts of a stream job.
+
+Let's consider a cluster with a single stream job, where this job has just 2 actors:
+```text
+Backfill (A) -> Stream Materialize (B)
+```
+
+If B is busy, it just won't request new Stream Messages (Barriers and StreamChunks) from A.
+This is called Backpressure. You can see that A's stream will be backpressured by B, i.e. no longer progress, until B is ready again.
+
+Now consider if B had the following sequence of events:
+```text
+1. Scheduled Barrier X (<1ms processing time)
+2. StreamChunk (250ms processing time)
+3. Scheduled Barrier Y (<1ms processing time)
+```
+
+The barrier latency for Y will be 0ms, since our barrier interval is 256ms,
+and the time between X and Y is less than 256ms.
+
+Consider another case where B has the following sequence of events:
+```text
+1. Scheduled Barrier X (<1ms processing time)
+2. StreamChunk (256ms processing time)
+3. StreamChunk (250ms processing time)
+4. Scheduled Barrier Y (<1ms processing time)
+```
+
+Now barrier latency will be around 250ms.
+
+In the case above, where we have Backfill and Stream Materialize, this is unlikely to occur however.
+Backfill should always let upstream side (barriers) bypass, so we should see stream chunk (3) only after (4).
+
+### Adding Channels to the equation
+
+Now consider a case where we have `join`.
+
+```
+StreamScan A -- HashJoin X
+              /
+             /
+StreamScan B
+```
+
+Now both A and B will have a channel between themselves and X.
+The channel is there to provide a buffer between actors, since A and B are not directly connected to X, they could be on different machines.
+Hence providing a buffer isolates X from network latency of getting Stream Messages from A and B.
+
+If X is busy, it will first backpressure the channel.
+However, the channel will not immediately backpressure A and B.
+It will only do so once it is full.
+
+This means that the stream chunks between barriers are equivalent to channel capacity.
+And then the barrier latency is the duration taken to process these barriers.
+
+Besides channels, some executors may also buffer chunks, for instance within the ChunkBuilder.
+Barrier latency also has to consider these.
+
+_example.toml_
+```toml
+[streaming.developer]
+stream_exchange_initial_permits = 2048
+```
+
+### Considering Amplification and UDF latency
+
+When there's high amplification or UDF latency in some downstream actor, upstream will naturally be backpressured.
```

Comment:
"Amplification" is not a widely-used term, although we are all know it. Let's add some introduction of it.

```
Amplification refers to a common phenomenon in streaming. For example, consider the following Join:

   SELECT * FROM users JOIN orders ON orders.user_id = users.id

We know that each user often has multiple orders, and a minority of users may have thousands of orders. If a field in the user table is updated, then reflected in the results of the above JOIN query, there could be thousands of changed rows. Such amplification can have unexpected performance impacts on streaming jobs, and therefore requires special attention.
```